### ISSUE TITLE
Is there a way to enter stopwords and / or c-TF-IDF after I optimised hbdscan with topic tuning?

### ISSUE THREAD

**Issue by chrislalk** [drob-xx/TopicTuner#9]
With BERTopic it is possible to select important words via c-TF-IDF and stopwords removal after the hdbscan calculation. Is this possible using TopicTuner as well? (Or would that lead to problems, because TopicTuner does not optimize for this post-processing state and then would cause much more uncategorized topics?).

In this case, would it make sense to do the stopwords removal and c-TF-IDF at the beginning before topic tuning?

Thanks for your consideration!!

**Comment by drob-xx** [comment]
Hi @chrislalk. The short answer is "not explicitly within TopicTuner". The longer answer is that you can embed any set of words you wish and then cluster that. So before you embed you could trim the input. However, following the pattern and approach followed in BERTopic I'm not sure what would be gained. The language models is created without any text pre-processing so theoretically (and practically from what I've seen) doing pre-processing would essentially just throw the embeddings off. Within BERTopic, cTF-IDF is run against the clusters created by HDBSCAN - so there is no opportunity to do that before clustering.

**Comment by chrislalk** [comment]
Thanks for your quick reply @drob-xx!! I see the problem that I would just trip off my language model if I did a pre-processing. However, I'm not sure where to go from here. Because I recently used TopicTuner successfully to create a topic model with zero uncategorized topics. But then I realized that my topics were mainly just consisting from stopwords, which wasn't really helpful. Is there a way out of this conundrum? Did you ever run into the same problem using TopicTuner? Or did I maybe just use too few documents, because I just selected 500 docs (and ended up with only two topics) from altogether over 100k docs?

**Comment by drob-xx** [comment]
np @chrislalk. So the question is why are there a lot of stopwords? FYI, just optimizing for few or no outliers is not necessarily a good idea. Either your text is very clearly segmented or you have chosen parameters that are only giving you a small number (or a single) cluster - probably not ideal. However, the fact that there are a lot of stopwords is not necessarily a problem. If you are happy with the clustering then I would simply export the BERTopic model and adjust the stopwords with BERTopic. Subsequent runs of BERTopic will maintain the HDBSCAN parameters you have chosen and the UMAP that is exported will use the same value for `random_state` which will mean that as long you don't switch architectures from the environment you used to create the model with TopicTuner, then you will get the same results as with TopicTuner.

**Comment by chrislalk** [comment]
Right! That makes sense. 

The way I did it now is even without TopicTuner export, just by using the HDBSCAN parameters from TopicTuner. 
```
ctfidf_model = ClassTfidfTransformer() # initiate ctfidf modell

# initiate hbdscan model with parameters resulting from topictuner
hdbscan_model = HDBSCAN(min_cluster_size=5, min_samples = 5, metric='euclidean', prediction_data=True) 

# putting it all together with addition parameters 
BERT_model = BERTopic(hdbscan_model=hdbscan_model, ctfidf_model=ctfidf_model,
                      calculate_probabilities = True, verbose = True, n_gram_range = (1,2))

topics, probs = BERT_model.fit_transform(tmt.docs, tmt.embeddings) # fitting the model
```
Does that make sense?
Thanks again!

**Comment by drob-xx** [comment]
I don't think that will work b/c you are not fixing the UMAP `random_state` which will effect the final params for HDBSCAN. You are also creating a ctfidf_model when you don't have to. TopicTuner was designed to make this process easy. When you generate a TopicTuner model it internally sets the `random_state` which will then be exported automatically when you create the BERTopic model with `.getBERTopicModel()`. So the workflow would be something like:

- create TMT model
- optimize and find your params
- optionally set `.bestParams` to those params. This is a convenience, bestParams will be used by a number of the procedures - including `.getBERTopicModel()` later if you don't specify the params.
- get a BERTopic model using `.getBERTopicModel()`
- create a CountVectorizer with the stop words you want to use
- Set the BERTopic model use the CountVectorizer instance you created
- Run `BERTopic.fit_transform`

```
from sklearn.feature_extraction import text
from sklearn.feature_extraction.text import CountVectorizer

tmt = TMT()
tmt.bestParams = (5,5) # This sets default parameters - not necessary, see next
BERTopic_model = tmt.getBERTopicModel() # this will use the bestParams, but you could also just getBERTopicModel(5,5)

#Now use the standard english stop words list and add any words you might want to stop
my_stop_words = text.ENGLISH_STOP_WORDS.union(['said', 'told', 'just', 'mr', 'new', 'like', 'use'])

# I like to use trigrams as well as bigrams - but it could be 1,2 etc.
cv = CountVectorizer(ngram_range=(1, 3), stop_words=my_stop_words)

BERTopic_model.vectorizer_model = cv
topics, probabilities = BERTopic_model.fit_transform(tmt.docs, embeddings=tmt.embeddings)
```
I haven't tested this exact code above - so you may have to play with it a bit.

**Comment by chrislalk** [comment]
Thank you very much! That looks great!
Will put it into practice as soon as I have my high-quality embeddings :)

**Comment by drob-xx** [comment]
Closed due to lack of activity.
